Cannot set single location properties when locations are already defined

### Jenkins and plugins versions report

<details>
  <summary>Environment</summary>

  <!-- Paste your environment details below -->
  ```text
Jenkins: 2.555.1
OS: Linux - 6.6.87.2-microsoft-standard-WSL2
Java: 21.0.10 - Eclipse Adoptium (OpenJDK 64-Bit Server VM)
---
analysis-model-api:14.1.0-954.v44a_a_5856e709
antisamy-markup-formatter:173.v680e3a_b_69ff3
apache-httpcomponents-client-4-api:4.5.14-269.vfa_2321039a_83
apache-httpcomponents-client-5-api:5.6-193.vf028a_770a_fa_c
asm-api:9.9.1-189.vb_5ef2964da_91
bootstrap5-api:5.3.8-1024.v127320880c60
bouncycastle-api:2.30.1.84-291.v9f17b_21896e2
branch-api:2.1280.v0d4e5b_b_460ef
caffeine-api:3.2.3-194.v31a_b_f7a_b_5a_81
checks-api:402.vca_263b_f200e3
cloudbees-folder:6.1100.ve9eed61d16c4
commons-lang3-api:3.20.0-109.ve43756e2d2b_4
commons-text-api:1.15.0-218.va_61573470393
credentials:1502.v5c95e620ddfe
credentials-binding:719.v80e905ef14eb_
data-tables-api:2.3.7-1554.v6113418a_6935
display-url-api:2.217.va_6b_de84cc74b_
durable-task:664.v2b_e7a_dfff66c
echarts-api:6.0.0-1273.v36fc9fe89332
font-awesome-api:7.2.0-983.v3f63c34eddb_9
forensics-api:4.1875.vefb_f97cb_4987
gson-api:2.13.2-198.v45e4a_55d9b_a_a_
instance-identity:203.v15e81a_1b_7a_38
ionicons-api:94.vcc3065403257
jackson2-api:2.21.2-433.v6d50b_92cfe52
jackson3-api:3.1.2-71.v0a_23916945fe
jakarta-activation-api:2.1.4-1
jakarta-mail-api:2.1.5-1
jakarta-xml-bind-api:4.0.6-12.vb_1833c1231d3
javax-activation-api:1.2.0-8
jaxb:2.3.9-143.v5979df3304e6
joda-time-api:2.14.1-187.vdf2def02b_8a_1
jquery3-api:3.7.1-682.vfa_cdce169929
json-api:20251224-185.v0cc18490c62c
json-path-api:3.0.0-218.vcd4dd1355de2
jsoup:1.22.1-76.v9cdb_2456c0e3
mailer:534.v1b_36f5864073
pipeline-build-step:584.vdb_a_2cc3a_d07a_
pipeline-groovy-lib:797.v90ea_a_9b_e45a_0
pipeline-input-step:551.vdff487c5998c
pipeline-milestone-step:152.v6e22b_8cfc66c
pipeline-model-api:2.2277.v00573e73ddf1
pipeline-model-definition:2.2277.v00573e73ddf1
pipeline-model-extensions:2.2277.v00573e73ddf1
pipeline-stage-step:345.va_96187909426
pipeline-stage-tags-metadata:2.2277.v00573e73ddf1
plain-credentials:199.v9f8e1f741799
plugin-util-api:7.1320.v684dd26fca_19
prism-api:1.30.0-720.v1eb_7496954b_3
scm-api:728.vc30dcf7a_0df5
script-security:1399.ve6a_66547f6e1
snakeyaml-api:2.5-149.v72471e9c6371
snakeyaml-engine-api:3.0.1-5.vd98ea_ff3b_92e
ssh-credentials:372.va_250881b_08cd
structs:362.va_b_695ef4fdf9
variant:70.va_d9f17f859e0
warnings-ng:12.10012.v753ff0d0363a_
woodstox-core-api:7.1.1-1.v4d297985f397
workflow-aggregator:608.v67378e9d3db_1
workflow-api:1413.v2ff1a_5e720fa_
workflow-basic-steps:1098.v808b_fd7f8cf4
workflow-cps:4285.v8df38f05c3c5
workflow-durable-task-step:1475.ved562f6ec8b_3
workflow-job:1571.vb_423c255d6d9
workflow-multibranch:821.vc3b_4ea_780798
workflow-scm-step:466.va_d69e602552b_
workflow-step-api:724.v538c2362b_dfb_
workflow-support:1015.v785e5a_b_b_8b_22

  ```

</details>


### What Operating System are you using (both controller, and any agents involved in the problem)?

jenkins/jenkins:2.555.1-lts

### Reproduction steps

Start jenkins in a docker container:
```
docker run -d -p 8080:8080 --name jenkins2551 -v jenkins2551_home:/var/jenkins_home jenkins/jenkins:2.555.1-lts
```

install the following plugins during the wizard:

- Warnings
- Pipeline

Create the following pipeline job:
```
pipeline {
    agent any

    stages {
        stage('build') {
            steps {
            writeFile file: 'build.log', text: '''/workspace/foo.cpp:134:26: warning: declaration of \'i\' shadows a previous local [-Wshadow]
  134 |                 for (u32 i = 0; i < (cmd.Size-sizeof(TDSPCommand))/4 && (cmd.Size>sizeof(TDSPCommand))>0; i++)
      |                          ^
/workspace/foo.cpp:122:18: note: shadowed declaration is here
  122 |         for (int i = 0; i < _commandsCount; i++)
      |                  ^'''
                recordIssues enabledForFailure: true,sourceCodeRetention: 'LAST_BUILD',tools: [gcc(pattern: 'build.log', id: 'arm-gcc', name: 'arm-gcc')]
            }
        }
    }
}
```

Run pipeline job and observe that the warnings parser could not parse the warning.

The warning written to the build.log file comes from an actual gcc cross compiler:
```
vscode ➜ /workspace (development) $ /opt/arm-chr2/bin/arm-none-linux-gnueabi-g++ --version
arm-none-linux-gnueabi-g++ (crosstool-NG 1.25.0_rc1.5_6ddf9e8) 11.2.0
```

This is the actual warning text from the compiler:
```
/workspace/foo.cpp:134:26: warning: declaration of \'i\' shadows a previous local [-Wshadow]
  134 |                 for (u32 i = 0; i < (cmd.Size-sizeof(TDSPCommand))/4 && (cmd.Size>sizeof(TDSPCommand))>0; i++)
      |                          ^
/workspace/foo.cpp:122:18: note: shadowed declaration is here
  122 |         for (int i = 0; i < _commandsCount; i++)
      |                  ^
```

### Expected Results

```
Started by user admin
[Pipeline] Start of Pipeline
[Pipeline] node
Running on Jenkins in /var/jenkins_home/workspace/test
[Pipeline] {
[Pipeline] stage
[Pipeline] { (build)
[Pipeline] writeFile
[Pipeline] recordIssues
[arm-gcc] Searching for all files in '/var/jenkins_home/workspace/test' that match the pattern 'build.log'
[arm-gcc] Traversing of symbolic links: enabled
[arm-gcc] -> found 1 file
[arm-gcc] Successfully parsed file /var/jenkins_home/workspace/test/build.log
[arm-gcc] -> found 1 issue (skipped 0 duplicates)
[arm-gcc] Successfully processed file 'build.log'
[arm-gcc] Post processing issues on 'Master' with source code encoding 'UTF-8'
[arm-gcc] Creating SCM blamer to obtain author and commit information for affected files
[arm-gcc] -> no SCM found
[arm-gcc] Resolving file names for all issues in workspace '/var/jenkins_home/workspace/test'
[arm-gcc] -> resolved paths in source directory (0 found, 1 not found)
[arm-gcc] Resolving module names from module definitions (build.xml, pom.xml, or Manifest.mf files)
[arm-gcc] -> resolved module names for 1 issues
[arm-gcc] Resolving package names (or namespaces) by parsing the affected files
[arm-gcc] -> resolved package names of 1 affected files
[arm-gcc] No filter has been set, publishing all 1 issues
[arm-gcc] Creating fingerprints for all affected code blocks to track issues over different builds
[arm-gcc] -> created fingerprints for 0 issues (skipped 1 issues)
[arm-gcc] Copying affected files to Jenkins' build folder '/var/jenkins_home/jobs/test/builds/8/files-with-issues'
[arm-gcc] -> 0 copied, 0 not in workspace, 1 not-found, 0 with I/O error
[arm-gcc] Deleting source code files of build jenkinsci/analysis-model-api-plugin#7
[arm-gcc] Repository miner is not configured, skipping repository mining
[arm-gcc] [-ERROR-] Can't create fingerprints for some files:
[arm-gcc] [-ERROR-] - '/workspace/foo.cpp' file not found
[arm-gcc] Reference build recorder is not configured
[arm-gcc] No valid reference build found
[arm-gcc] All reported issues will be considered outstanding
[arm-gcc] No quality gates have been set - skipping
[arm-gcc] Health report is disabled - skipping
[arm-gcc] Created analysis result for 1 issues (found 0 new issues, fixed 0 issues)
[arm-gcc] Attaching ResultAction with ID 'arm-gcc' to build 'test jenkinsci/analysis-model-api-plugin#8'.
[Checks API] No suitable checks publisher found.
[Pipeline] }
[Pipeline] // stage
[Pipeline] }
[Pipeline] // node
[Pipeline] End of Pipeline
Finished: SUCCESS
```

### Actual Results

```
Started by user admin
[Pipeline] Start of Pipeline
[Pipeline] node
Running on Jenkins in /var/jenkins_home/workspace/test
[Pipeline] {
[Pipeline] stage
[Pipeline] { (build)
[Pipeline] writeFile
[Pipeline] recordIssues
[Pipeline] }
[Pipeline] // stage
[Pipeline] }
[Pipeline] // node
[Pipeline] End of Pipeline
Also:   org.jenkinsci.plugins.workflow.actions.ErrorAction$ErrorId: db08e196-49a5-4659-a0a4-47af86d1a8aa
java.lang.AssertionError: Cannot set single location properties when locations are already defined
	at PluginClassLoader for plugin-util-api//edu.hm.hafner.util.Ensure.throwException(Ensure.java:172)
	at PluginClassLoader for plugin-util-api//edu.hm.hafner.util.Ensure$BooleanCondition.isTrue(Ensure.java:724)
	at PluginClassLoader for analysis-model-api//edu.hm.hafner.analysis.IssueBuilder.ensureThatLocationsAreEmpty(IssueBuilder.java:139)
	at PluginClassLoader for analysis-model-api//edu.hm.hafner.analysis.IssueBuilder.setFileName(IssueBuilder.java:154)
	at PluginClassLoader for analysis-model-api//edu.hm.hafner.analysis.parser.Gcc4CompilerParser.createIssue(Gcc4CompilerParser.java:91)
	at PluginClassLoader for analysis-model-api//edu.hm.hafner.analysis.LookaheadParser.parse(LookaheadParser.java:82)
	at PluginClassLoader for analysis-model-api//edu.hm.hafner.analysis.LookaheadParser.parseReport(LookaheadParser.java:65)
	at PluginClassLoader for analysis-model-api//edu.hm.hafner.analysis.IssueParser.parse(IssueParser.java:69)
	at PluginClassLoader for analysis-model-api//edu.hm.hafner.analysis.registry.CompositeParserDescriptor$CompositeParser.parseReport(CompositeParserDescriptor.java:90)
	at PluginClassLoader for analysis-model-api//edu.hm.hafner.analysis.IssueParser.parse(IssueParser.java:69)
	at PluginClassLoader for warnings-ng//io.jenkins.plugins.analysis.core.model.IssueReportScanner.processFile(IssueReportScanner.java:54)
	at PluginClassLoader for plugin-util-api//io.jenkins.plugins.util.AgentFileVisitor.scanFiles(AgentFileVisitor.java:116)
	at PluginClassLoader for plugin-util-api//io.jenkins.plugins.util.AgentFileVisitor.invoke(AgentFileVisitor.java:95)
	at PluginClassLoader for plugin-util-api//io.jenkins.plugins.util.AgentFileVisitor.invoke(AgentFileVisitor.java:40)
	at hudson.FilePath.act(FilePath.java:1223)
	at hudson.FilePath.act(FilePath.java:1206)
	at PluginClassLoader for warnings-ng//io.jenkins.plugins.analysis.core.model.ReportScanningTool.scanInWorkspace(ReportScanningTool.java:193)
	at PluginClassLoader for warnings-ng//io.jenkins.plugins.analysis.core.model.ReportScanningTool.scan(ReportScanningTool.java:172)
	at PluginClassLoader for warnings-ng//io.jenkins.plugins.analysis.core.model.ReportScanningTool.scan(ReportScanningTool.java:156)
	at PluginClassLoader for warnings-ng//io.jenkins.plugins.analysis.core.steps.IssuesScanner.scan(IssuesScanner.java:117)
	at PluginClassLoader for warnings-ng//io.jenkins.plugins.analysis.core.steps.IssuesRecorder.scanWithTool(IssuesRecorder.java:835)
	at PluginClassLoader for warnings-ng//io.jenkins.plugins.analysis.core.steps.IssuesRecorder.record(IssuesRecorder.java:750)
	at PluginClassLoader for warnings-ng//io.jenkins.plugins.analysis.core.steps.IssuesRecorder.perform(IssuesRecorder.java:726)
	at PluginClassLoader for warnings-ng//io.jenkins.plugins.analysis.core.steps.RecordIssuesStep$Execution.run(RecordIssuesStep.java:692)
	at PluginClassLoader for warnings-ng//io.jenkins.plugins.analysis.core.steps.RecordIssuesStep$Execution.run(RecordIssuesStep.java:647)
	at PluginClassLoader for workflow-step-api//org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution.lambda$start$0(SynchronousNonBlockingStepExecution.java:49)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
Finished: FAILURE
```

### Anything else?

This issue does not happen with the Analysis Model API version:
[13.21.0-945.va_f5a_0b_c1404b](https://updates.jenkins.io/download/plugins/analysis-model-api/13.21.0-945.va_f5a_0b_c1404b_/analysis-model-api.hpi)

### Are you interested in contributing a fix?

_No response_

--- a/src/main/java/edu/hm/hafner/analysis/IssueBuilder.java
+++ b/src/main/java/edu/hm/hafner/analysis/IssueBuilder.java
@@ -500,6 +500,19 @@ public class IssueBuilder implements AutoCloseable {
      * @return this
      */
     @CanIgnoreReturnValue
+    public IssueBuilder setMessage(final StringBuilder message) {
+        return setMessage(message.toString());
+    }
+
+    /**
+     * Sets the detailed message for this issue.
+     *
+     * @param message
+     *         the message
+     *
+     * @return this
+     */
+    @CanIgnoreReturnValue
     public IssueBuilder setMessage(@CheckForNull final String message) {
         if (StringUtils.isBlank(message)) {
             this.message = EMPTY_TREE_STRING;
@@ -586,6 +599,30 @@ public class IssueBuilder implements AutoCloseable {
         }
 
         return this;
+    }
+
+    /**
+     * Adds another location to this issue, the first location is considered the primary location.
+     *
+     * @param fileName
+     *         the name of the affected file
+     * @param lineStart
+     *         the first line of the affected code
+     * @param lineEnd
+     *         the last line of the affected code
+     * @param columnStart
+     *         the first column of the affected code
+     * @param columnEnd
+     *         the last column of the affected code
+     *
+     * @return this
+     */
+    @CanIgnoreReturnValue
+    @SuppressWarnings("checkstyle:HiddenField")
+    public IssueBuilder addLocation(final String fileName,
+            final int lineStart, final int lineEnd,
+            final int columnStart, final int columnEnd) {
+        return addLocation(new Location(internFileName(fileName), lineStart, lineEnd, columnStart, columnEnd));
     }
 
     /**

--- a/src/main/java/edu/hm/hafner/analysis/parser/Gcc4CompilerParser.java
+++ b/src/main/java/edu/hm/hafner/analysis/parser/Gcc4CompilerParser.java
@@ -1,17 +1,16 @@
 package edu.hm.hafner.analysis.parser;
 
 import org.apache.commons.lang3.Strings;
+import org.dom4j.util.StringUtils;
 
 import edu.hm.hafner.analysis.Issue;
 import edu.hm.hafner.analysis.IssueBuilder;
-import edu.hm.hafner.analysis.Location;
 import edu.hm.hafner.analysis.LookaheadParser;
-import edu.hm.hafner.analysis.Severity;
 import edu.hm.hafner.analysis.util.IntegerParser;
 import edu.hm.hafner.util.LookaheadStream;
-import edu.hm.hafner.util.TreeStringBuilder;
 
 import java.io.Serial;
+import java.util.ArrayList;
 import java.util.Optional;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
@@ -25,12 +24,13 @@ public class Gcc4CompilerParser extends LookaheadParser {
     @Serial
     private static final long serialVersionUID = 5490211629355204910L;
 
+    private static final String LOCATION = "(?<file>.+?):(?<line>\\d+):(?:(?<column>\\d+):)?";
     private static final String GCC_WARNING_PATTERN =
-            ANT_TASK + "(.+?):(\\d+):(?:(\\d+):)? ?([wW]arning|.*[Ee]rror|[Nn]ote): (.*)$";
+            ANT_TASK + LOCATION + " ?(?<severity>[wW]arning|.*[Ee]rror): (?<message>.*)$";
     private static final Pattern CLASS_PATTERN = Pattern.compile("\\[-W(.+)]$");
     private static final Pattern GCC_OPTION_PATTERN = Pattern.compile("\\[-(f[^\\]]+)]$");
     private static final Pattern CLANG_TIDY_PATTERN = Pattern.compile("\\[[^\\s\\]]+\\]$");
-    private static final Pattern NOTE_PATTERN = Pattern.compile("^(?<file>.+?):(?<line>\\d+):(?:(?<column>\\d+):)? ?[Nn]ote: (?<message>.*)$");
+    private static final Pattern NOTE_PATTERN = Pattern.compile("^" + LOCATION + " ?[Nn]ote: (?<message>.*)$");
 
     /**
      * Creates a new instance of {@link Gcc4CompilerParser}.
@@ -40,8 +40,10 @@ public class Gcc4CompilerParser extends LookaheadParser {
     }
 
     /**
-     * Creates a new instance of {@link Gcc4CompilerParser} with specified pattern.
-     * @param pattern a regex pattern to be used instead of the default one
+     * Creates a new instance of {@link Gcc4CompilerParser} with the specified pattern.
+     *
+     * @param pattern
+     *         a regex pattern to be used instead of the default one
      */
     Gcc4CompilerParser(final String pattern) {
         super(pattern);
@@ -55,64 +57,72 @@ public class Gcc4CompilerParser extends LookaheadParser {
     @Override
     protected Optional<Issue> createIssue(final Matcher matcher, final LookaheadStream lookahead,
             final IssueBuilder builder) {
-        if (isNoteMessage(matcher)) {
+        var originalMessage = matcher.group("message");
+        if (isClangTidyWarning(originalMessage)) {
             return Optional.empty();
         }
 
-        var message = new StringBuilder(matcher.group(5));
-        var originalMessage = message.toString();
+        addLocation(matcher, builder);
 
-        setCategory(builder, originalMessage);
-
+        var message = new StringBuilder(originalMessage);
         boolean hasCodeSnippet = false;
         while (lookahead.hasNext() && isMessageContinuation(lookahead, hasCodeSnippet)) {
             var continuation = lookahead.next();
-            if (continuation.length() > 0 && Character.isWhitespace(continuation.charAt(0))) {
+            if (StringUtils.startsWithWhitespace(continuation)) {
                 hasCodeSnippet = true;
             }
             message.append('\n');
             message.append(continuation);
         }
 
-        var treeStringBuilder = new TreeStringBuilder();
-        var notes = collectNotes(lookahead, builder, treeStringBuilder);
+        var notes = collectNotes(lookahead, builder);
         if (!notes.isEmpty()) {
             message.append('\n');
             message.append(notes);
         }
 
-        // Reject clang-tidy warnings that have [check-name] but not [-Wwarning-name]
-        // clang-tidy warnings should be handled by ClangTidyParser
-        var messageStr = message.toString();
-        if (CLANG_TIDY_PATTERN.matcher(messageStr).find() && !CLASS_PATTERN.matcher(messageStr).find()) {
-            return Optional.empty();
-        }
-
-        return builder.setFileName(matcher.group(1))
-                .setLineStart(matcher.group(2))
-                .setColumnStart(matcher.group(3))
-                .setMessage(messageStr)
-                .setSeverity(Severity.guessFromString(matcher.group(4)))
+        setCategory(builder, originalMessage);
+        return builder.setMessage(message)
+                .guessSeverity(matcher.group("severity"))
                 .buildOptional();
     }
 
-    /**
-     * Checks if the matched line is a note message.
-     *
-     * @param matcher the matcher for the current line
-     * @return true if the line is a note message, false otherwise
-     */
-    private boolean isNoteMessage(final Matcher matcher) {
-        var messageType = matcher.group(4);
-        return "note".equals(messageType) || "Note".equals(messageType);
+    private void addLocation(final Matcher matcher, final IssueBuilder builder) {
+        var fileName = matcher.group("file");
+        var line = toInt(matcher.group("line"));
+        var column = toInt(matcher.group("column"));
+
+        builder.addLocation(fileName, line, line, column, column);
+    }
+
+    private int toInt(final String number) {
+        return IntegerParser.parseInt(number);
     }
 
     /**
-     * Sets the category based on the original message content.
-     * Checks for GCC warning categories (e.g., [-Wunused-variable]) or compiler options (e.g., [-fpermissive]).
+     * Check if the warning is from clang-tidy. Those warnings are quite similar and have [check-name] but not
+     * [-Wwarning-name] or [-fcompiler-option]. Since a separate parser handles clang-tidy warnings, we can skip them
+     * here.
      *
-     * @param builder the issue builder
-     * @param originalMessage the original message to extract the category from
+     * @param message
+     *         the original message to check
+     *
+     * @return {@code true} if the message is a clang-tidy warning, {@code false} otherwise
+     */
+    private boolean isClangTidyWarning(final String message) {
+        return CLANG_TIDY_PATTERN.matcher(message).find()
+                && !CLASS_PATTERN.matcher(message).find()
+                && !GCC_OPTION_PATTERN.matcher(message).find();
+    }
+
+    /**
+     * Sets the category based on the original message content. Checks for GCC warning categories (e.g.,
+     * [-Wunused-variable]) or compiler options (e.g., [-fpermissive]).
+     *
+     * @param builder
+     *         the issue builder
+     * @param originalMessage
+     *         the original message to extract the category from
      */
     private void setCategory(final IssueBuilder builder, final String originalMessage) {
         var classMatcher = CLASS_PATTERN.matcher(originalMessage);
@@ -128,38 +138,28 @@ public class Gcc4CompilerParser extends LookaheadParser {
     }
 
     /**
-     * Collects any subsequent note lines from the lookahead stream and adds them as additional locations.
+     * Collects any further note lines from the lookahead stream and adds them as additional locations.
      *
-     * @param lookahead the lookahead stream
-     * @param builder the issue builder to add locations to
-     * @param treeStringBuilder the tree string builder for interning file names
+     * @param lookahead
+     *         the lookahead stream
+     * @param builder
+     *         the issue builder to add locations to
+     *
      * @return a string containing all collected note lines, or an empty string if no notes were found
      */
-    private String collectNotes(final LookaheadStream lookahead, final IssueBuilder builder,
-            final TreeStringBuilder treeStringBuilder) {
-        var notes = new StringBuilder();
+    private String collectNotes(final LookaheadStream lookahead, final IssueBuilder builder) {
+        var notes = new ArrayList<String>();
         while (lookahead.hasNext()) {
-            var nextLine = lookahead.peekNext();
-            var noteMatcher = NOTE_PATTERN.matcher(nextLine);
+            var noteMatcher = NOTE_PATTERN.matcher(lookahead.peekNext());
             if (noteMatcher.matches()) {
-                if (!notes.isEmpty()) {
-                    notes.append('\n');
-                }
-                notes.append(lookahead.next());
-
-                // Add the note as an additional location
-                var file = noteMatcher.group("file");
-                var line = IntegerParser.parseInt(noteMatcher.group("line"));
-                var column = IntegerParser.parseInt(noteMatcher.group("column"));
-
-                var fileName = treeStringBuilder.intern(file);
-                builder.addLocation(new Location(fileName, line, line, column, column));
+                notes.add(lookahead.next());
+                addLocation(noteMatcher, builder);
             }
             else {
                 break;
             }
         }
-        return notes.toString();
+        return String.join("\n", notes);
     }
 
     @SuppressWarnings("PMD.AvoidLiteralsInIfCondition")

--- a/src/test/java/edu/hm/hafner/analysis/parser/Gcc4CompilerParserTest.java
+++ b/src/test/java/edu/hm/hafner/analysis/parser/Gcc4CompilerParserTest.java
@@ -622,11 +622,40 @@ class Gcc4CompilerParserTest extends AbstractParserTest {
         }
     }
 
-    /**
-     * Parses a javac warning log.
-     *
-     * @see <a href="https://issues.jenkins.io/browse/JENKINS-68396">Issue 68396</a>
-     */
+    @Test
+    @org.junitpioneer.jupiter.Issue("https://github.com/jenkinsci/analysis-model/issues/1488")
+    void shouldRegisterPrimaryLocationInTheBeginning() {
+        var report = parseStringContent(""" 
+                /workspace/foo.cpp:134:26: warning: declaration of \\'i\\' shadows a previous local [-Wshadow]
+                  134 |                 for (u32 i = 0; i < (cmd.Size-sizeof(TDSPCommand))/4 && (cmd.Size>sizeof(TDSPCommand))>0; i++)
+                      |                          ^
+                /workspace/foo.cpp:122:18: note: shadowed declaration is here
+                  122 |         for (int i = 0; i < _commandsCount; i++)
+                      |                  ^
+                """);
+
+        assertThat(report).hasSize(1);
+
+        var issue = report.get(0);
+        assertThat(issue)
+                .hasLineStart(134)
+                .hasColumnStart(26)
+                .hasFileName("/workspace/foo.cpp")
+                .hasSeverity(Severity.WARNING_NORMAL);
+        assertThat(issue.getMessage()).containsIgnoringWhitespaces("""
+                declaration of \\'i\\' shadows a previous local [-Wshadow]
+                       134 |                 for (u32 i = 0; i < (cmd.Size-sizeof(TDSPCommand))/4 && (cmd.Size>sizeof(TDSPCommand))>0; i++)
+                           |                          ^
+                     /workspace/foo.cpp:122:18: note: shadowed declaration is here
+                """);
+        assertThat(issue.getLocations()).hasSize(2);
+        assertThat(issue.getLocations().getLast())
+                .hasFileName("/workspace/foo.cpp")
+                .hasLineStart(122)
+                .hasColumnStart(18);
+    }
+
+    @org.junitpioneer.jupiter.Issue("https://issues.jenkins.io/browse/JENKINS-68396")
     @Test
     void issue68396() {
         var warnings = parse("issue68396.txt");

